boxoffice/forms/discount.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+
+from flask import g
+from baseframe import __
+import baseframe.forms as forms
+from baseframe.forms.sqlalchemy import QuerySelectMultipleField, QuerySelectField
+from coaster.utils import getbool
+from ..models import DISCOUNT_TYPE, DiscountPolicy, CURRENCY
+
+__all__ = ['DiscountPolicyForm', 'DiscountPriceForm', 'DiscountCouponForm']
+
+
+class DiscountPolicyForm(forms.Form):
+    title = forms.StringField(__("Discount Title"),
+        validators=[forms.validators.DataRequired((u"Please specify a discount title")),
+        forms.validators.Length(max=250), forms.validators.StripWhitespace()])
+    is_price_based = forms.RadioField(__("Price based discount"),
+        choices=[(True, 'Special price discount'),
+        (False, 'Percentage based discount')], coerce=getbool, default=True)
+    discount_type = forms.RadioField(__("Discount Type"),
+        choices=DISCOUNT_TYPE.items(), coerce=int, default=1)
+    percentage = forms.IntegerField(__("Percentage"),
+        validators=[forms.validators.OptionalIf('is_price_based', (u"Please specify a discount percentage"))])
+    item_quantity_min = forms.IntegerField(__("Minimum number of tickets"), default=1)
+    discount_code_base = forms.StringField(__("Discount Title"),
+        validators=[forms.validators.OptionalIfNot('discount_type',
+            (u"Please specify a discount code base")),
+        forms.validators.Length(max=20), forms.validators.StripWhitespace()], filters=[lambda code_base: code_base or None])
+    bulk_coupon_usage_limit = forms.IntegerField(__("Number of times a bulk coupon can be used"), default=1)
+    items = QuerySelectMultipleField(__("Items"), get_label='title', query_factory=lambda: [],
+        validators=[forms.validators.OptionalIf('is_price_based', (u"Please select a item to which discount is to be applied"))])
+    # For validate_discount_code_base
+    discount_policy_id = forms.StringField(__("Id"),
+        validators=[forms.validators.Optional()], default=None)
+
+    def validate_discount_code_base(self, field):
+        if self.discount_policy_id.data:
+            if DiscountPolicy.query.filter(DiscountPolicy.id != self.discount_policy_id.data, DiscountPolicy.discount_code_base == field.data).notempty():
+                raise forms.ValidationError((u"Please specify a different discount code base"))
+        elif DiscountPolicy.query.filter_by(discount_code_base=field.data).notempty():
```

Comment:
This is because discount policies were designed to work for items across item collections. Although, in hindsight, `discount_code_base` should perhaps have been designed to be unique per organization.